lint/src/main/kotlin/com/google/samples/apps/nowinandroid/lint/TestMethodNameDetector.kt

```diff
@@ -0,0 +1,160 @@
+        annotationInfo: AnnotationInfo,
+        usageInfo: AnnotationUsageInfo,
+    ) {
+        val method = usageInfo.referenced as? PsiMethod ?: return
+
+        method.detectPrefix(context, usageInfo)
+        method.detectFormat(context, usageInfo)
+        method.detectUnderscore(context, usageInfo)
+    }
+
+    private fun JavaContext.isAndroidTest() = Path("androidTest") in file.toPath()
+
+    private fun PsiMethod.detectPrefix(
+        context: JavaContext,
+        usageInfo: AnnotationUsageInfo,
+    ) {
+        if (!name.startsWith("test")) return
+        context.report(
+            issue = PREFIX,
+            scope = usageInfo.usage,
+            location = context.getNameLocation(this),
+            message = PREFIX.getBriefDescription(RAW),
+            quickfixData = LintFix.create()
+                .name("Remove prefix")
+                .replace().pattern("""test[\s_]*""")
+                .with("")
+                .autoFix()
+                .build(),
+        )
+    }
+
+    private fun PsiMethod.detectFormat(
+        context: JavaContext,
+        usageInfo: AnnotationUsageInfo,
+    ) {
+        if (!context.isAndroidTest()) return
+        if ("""[^\W_]+(_[^\W_]+){1,2}""".toRegex().matches(name)) return
+        context.report(
+            issue = FORMAT,
+            scope = usageInfo.usage,
+            location = context.getNameLocation(this),
+            message = FORMAT.getBriefDescription(RAW),
+        )
+    }
+
+    private fun PsiMethod.detectUnderscore(
+        context: JavaContext,
+        usageInfo: AnnotationUsageInfo,
+    ) {
+        if (context.isAndroidTest()) return
+        if ("_" !in name) return
+        context.report(
+            issue = UNDERSCORE,
+            scope = usageInfo.usage,
+            location = context.getNameLocation(this),
+            message = UNDERSCORE.getBriefDescription(RAW),
+            quickfixData = LintFix.create()
+                .name("Replace underscores with spaces")
+                .replace()
+                .range(context.getNameLocation(this))
+                .with(
+                    name.replace("_", " ")
+                        .removeSurrounding("`")
+                        .let { """`$it`""" },
+                )
+                .autoFix()
+                .build(),
+        )
+    }
+
+    companion object {
+
+        private fun issue(
+            id: String,
+            briefDescription: String,
+            explanation: String,
+        ): Issue = Issue.create(
+            id = id,
+            briefDescription = briefDescription,
+            explanation = explanation,
+            category = TESTING,
+            priority = 5,
+            severity = WARNING,
+            implementation = Implementation(
+                TestMethodNameDetector::class.java,
+                EnumSet.of(JAVA_FILE, TEST_SOURCES),
+            ),
+        )
+
+        @JvmField
+        val UNDERSCORE: Issue = issue(
```

Comment:
The `UNDERSCORE` check is incompatible with the `FORMAT` check since the method name must contain 1-2 underscores. Suggest removing the `UNDERSCORE` check entirely.

Reply:
I submitted this change already.

Reply:
This was on purpose, the `UNDERSCORE` only applies to non-Android tests, where we should use backticks instead of undescores. 🙈

Reply:
@dturner Unless you really want to force this ugly snake_case format on regular jvm unit tests.
To me, the underscores workaround was only there for Android instrumented tests that do not support (for whatever reason) spaces in their names, and to make these test names more consistent and readable.

Reply:
I see, sorry, the intention wasn't clear from the PR description. I'd prefer to have a single convention, even if it's ugly.

Reply:
Ok